### PR TITLE
enhancement(demo_logs source): comply with component spec

### LINKED ISSUE #10065: Instrument the `demo_logs` (né `generator`) source with event processing metrics
Per #9326

### PR DESCRIPTION
**Pull request by jdrouet**
Add `BytesReceived`, `EventsReceived` and `StreamClosedError`.

closes #10065

### CODE REVIEW

File: src/internal_events/common.rs [hunk 1]
```diff
@@ -90,3 +90,36 @@ impl InternalEvent for AwsBytesSent {
         );
     }
 }
+
+#[derive(Debug)]
+pub struct StreamClosedError {
```
Comment (001wwang) [inline src/internal_events/common.rs:95]: Looking forward to when this is finally included in `master` 😄

File: src/sources/demo_logs.rs [hunk 2]
```diff
@@ -171,11 +179,12 @@ async fn demo_logs_source(
                         log.try_insert(log_schema().source_type_key(), Bytes::from("demo_logs"));
                         log.try_insert(log_schema().timestamp_key(), now);
 
-                        out.send(event)
-                            .await
-                            .map_err(|_: crate::source_sender::ClosedError| {
+                        out.send(event).await.map_err(
+                            |error: crate::source_sender::ClosedError| {
+                                emit!(&StreamClosedError { error, count: 1 });
```
Comment (001wwang) [inline src/sources/demo_logs.rs:184]: Is it worth updating the `count` to be the remaining number of events rather than `1`?

**Review by 001wwang — APPROVED** [review]
LGTM! Just one non-blocking suggestion.

**Comment by 001wwang** [comment]
Looks like `demo_logs source` isn't defined in https://github.com/vectordotdev/vector/blob/master/.github/semantic.yml

File: src/sources/demo_logs.rs (hunk 2 above)
Comment (jdrouet) [reply src/sources/demo_logs.rs:184]: Oh, I see your point! We could also do a `send_all` instead of sending them one by one 😅

**Comment by jdrouet** [comment]
@001wwang if you're ok with my changes (streaming events instead of sending them one by one), feel free to trigger the merge 😉

File: src/sources/demo_logs.rs [hunk 3]
```diff
@@ -156,27 +156,32 @@ async fn demo_logs_source(
...
 
         let line = format.generate_line(n);
 
         let mut stream = FramedRead::new(line.as_bytes(), decoder.clone());
         while let Some(next) = stream.next().await {
             match next {
-                Ok((events, _byte_size)) => {
+                Ok((events, byte_size)) => {
+                    let count = events.len();
+                    emit!(&EventsReceived { count, byte_size });
```
Comment (jszwedko) [inline src/sources/demo_logs.rs:171]: This should also be using `BytesSizeOf::size_of` rather than the raw frame byte count.

**Pull request merged**